WordPress/gutenberg
# Experimental: Navigation block with the Interactivity API

## What?

This pull request adds an experimental option to toggle between the Navigation block using [the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) and the old scripts that use [`micromodal`](https://micromodal.vercel.app/):

* Option deactivated: The navigation block keeps working exactly as it currently does.
* Option activated: The JS files are substituted for the Interactivity API logic, and it adds the appropriate directives to the markup using the HTML Tag Processor.

I recorded this video explaining in detail the purpose of the pull request and how it has been implemented:

https://www.loom.com/share/c07afacb4e464de8a8e7b253930dcd1b

## Why?

This pull request is part of the [proposal to use the Interactivity API](https://make.wordpress.org/core/2023/03/30/proposal-the-interactivity-api-a-better-developer-experience-in-building-interactive-blocks/) as a standard for the blocks frontend. The main goal is to start testing real blocks with this API, which could also serve as an example, and see what they would look like.

In terms of performance, it is true that, if you only use the Interactivity API for the navigation block, it sends a few more kBs to the client than the current approach (12kBs vs 4kBs). However, if more blocks use the Interactivity API, they would reuse those 12kBs and would result in fewer kBs overall. There are already ongoing pull requests that aim to use this code like the [Lightbox in images](https://github.com/WordPress/gutenberg/pull/49621). And there could be more core blocks in the future using it.

Additionally, this would prevent the block from depending on external libraries like `micromodal`.

## How?

* It creates an experimental setting to toggle between the Navigation block using the Interactivity API and the old scripts (micromodal).
* Depending on that setting, it enqueues the relevant scripts and adds the Interactivity API directives to the markup using the Tag Processor. If we finally decide to use the Interactivity API for the navigation block, we can move the directives directly to the markup.
* It adds the logic needed to replicate what `micromodal` is doing.

There are several options in the navigation block. These are the use cases taken into account to replicate the current behavior:

1. **Overlay menu**: The menu modal that opens after clicking the hamburger button.
![Screenshot 2023-05-09 at 10 03 24](https://user-images.githubusercontent.com/34552881/237033562-4430e89d-f4b5-4592-95c1-72defcea8823.png)

* Overlay menu opens on click on "open menu" button.
* Overlay menu closes on click on "close menu" button.
* Overlay menu closes when ESC key is pressed.
* Overlay menu focuses on first element after opening.
* Overlay menu traps focus.

2. **Submenus (open on click)**: Submenus when "Open on click" option is activated:
![Screenshot 2023-05-09 at 10 06 32](https://user-images.githubusercontent.com/34552881/237034241-914d35b2-117f-4134-89f4-2401c18c6cca.png)

* Submenu opens on click.
* Nested submenu opens on click.
* Submenu closes on click outside.
* Nested submenu closes on click outside.
* Submenu closes on ESC key.
* Nested submenu closes on ESC key
* Submenu closes on tab outside submenu.
* Only nested submenu closes on tab outside.

3. **Submenus (Arrow setting)**: Submenus when "Open on click" is deactivated and "Show arrow" is activated.
![Screenshot 2023-05-09 at 10 11 22](https://user-images.githubusercontent.com/34552881/237035480-f42ed39e-c663-4d8b-b71c-0ed77587b1ad.png)

* Submenu opens on click in the arrow. The text is currently a link.
* The rest of the aspects should behave the same way as with the other option.

4. **Page list block**: The Page list block, which should behave as the submenus.

I tried to cover all the use cases, but let me know if you are missing anything.

## Testing

Basically, we have to ensure that the Navigation block, the navigation submenu, and the page list behave exactly the same when the option is activated/deactivated.

* Toggle the Experimental option in Gutenberg -> Experiments:
  * Check that the mentioned experiences (and anything not mentioned) work as expected.
  * Check that it is only sending the Interactivity API scripts.
  * Change the settings of the navigation block to ensure it works fine in all cases.
  * If the Overlay menu is set to off and the "Show arrow" is set to false, the scripts shouldn't be loaded.

As part of this pull request, I added [e2e tests for the navigation](https://github.com/WordPress/gutenberg/pull/50462) to check all those user experiences mentioned previously. With both options tests are passing, but maybe there are more use cases not contemplated there.
